lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -0,0 +1,322 @@
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/TinyPtrVector.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#define DEBUG_TYPE "gct"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Traits
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+template <>
+struct DenseMapInfo<StringAttr> {
+  static inline StringAttr getEmptyKey() {
+    auto pointer = DenseMapInfo<void *>::getEmptyKey();
+    return StringAttr(static_cast<Attribute::ImplType *>(pointer));
+  }
+  static inline StringAttr getTombstoneKey() {
+    auto pointer = DenseMapInfo<void *>::getTombstoneKey();
+    return StringAttr(static_cast<Attribute::ImplType *>(pointer));
+  }
+  static unsigned getHashValue(const StringAttr &val) {
+    return mlir::hash_value(val);
+  }
+  static bool isEqual(const StringAttr &lhs, const StringAttr &rhs) {
+    return lhs == rhs;
+  }
+};
+} // end namespace llvm
+
+//===----------------------------------------------------------------------===//
+// Static class names
+//===----------------------------------------------------------------------===//
+
+static constexpr const char *signalDriverAnnoClass =
+    "sifive.enterprise.grandcentral.SignalDriverAnnotation";
+
+//===----------------------------------------------------------------------===//
+// Per-Module Signal Mappings
+//===----------------------------------------------------------------------===//
+
+namespace {
+enum class MappingDirection {
+  DriveRemote,
+  ProbeRemote,
+};
+
+struct SignalMapping {
+  MappingDirection dir;
+  StringAttr remoteTarget;
+  StringRef localName;
+  StringRef localFields;
+
+  /// The type of the signal being mapped.
+  FIRRTLType type;
+  /// The `localName` and `localFields` resolved to an actual value.
+  Value localValue;
+};
+
+struct ModuleSignalMappings {
+  ModuleSignalMappings(FModuleOp module) : module(module) {}
+  void run();
+  void addTargets(ArrayAttr targets, MappingDirection dir);
+  void emitMappingsModule();
+  void instantiateMappingsModule();
+
+  FModuleOp module;
+  bool anyFailed = false;
+  SmallVector<SignalMapping> mappings;
+  SmallString<64> mappingsModuleName;
+
+  /// A lookup table of local operations that carry a name which can be
+  /// referenced in an annotation.
+  DenseMap<StringRef, Value> localNames;
```

Comment:
As far as I understand, `StringMap` allocates the string keys locally so they become stable across modifications. That's not needed here, since the strings all come from `StringAttr`s already interned in the `MLIRContext`.

Reply:
I'd recommend making this attribute based, `DenseMap<Attribute, Value>` so you're just hashing the pointer in an attribute.  You can use `op.nameAttr()` to get it.

Reply:
Similarly, plz store the attribute (not the StringRef) in SignalMapping